sampleapp/src/main/scala/de/up/hpi/informationsystems/sampleapp/dactors/Cart.scala

```diff
@@ -54,8 +65,123 @@ class Cart(id: Int) extends Dactor(id) {
+    case AddItems.Request(orders, customerId) =>
+      addItems(orders, customerId).map(res => res match {
+        case Success(newRecords) => sender() ! AddItems.Success(newRecords.head.get(CartPurchases.sessionId).get)
+        case Failure (e) => sender() ! AddItems.Failure(e)
+      })
+
     case Checkout.Request(_) => sender() ! Checkout.Failure(new NotImplementedError)
   }
 
+  def addItems(orders: Seq[AddItems.Order], customerId: Int): Future[Try[Seq[Record]]] = {
+    // ask for prices of orders
+    // i_id, i_price, i_min_price
+    val prices = Future.sequence(orders
+      .groupBy(order => order.sectionId)
+      .map(askStoreSectionForPrices))
+
+    // c_g_id
+    val customer_group_id = askCustomerForGroupId(customerId)
+
+    // collect i_ids, group_id to query for fixed_disc
+    val i_ids_and_group_id = for {
+      prices <- prices
+      c_g_id <- customer_group_id
+    } yield (prices.flatten.map(ppr => ppr.i_id), c_g_id)
+
+    // i_id, fixed_disc
+    val fixed_disc = i_ids_and_group_id.flatMap(tuple => {
+      val (i_ids, group_id) = tuple
+      askGroupManagerForDiscount(i_ids.toSeq, group_id)
+    })
+
+    // join prices with their discounts:
+    // sec_id, i_id, sess_id (new), i_quantity, i_fixed_disc, i_min_price, i_price
+    val prices_with_discounts = for {
+      prices <- prices
+      fixed_disc <- fixed_disc
+    } yield joinPriceAndDisc(prices.flatten.toSeq, fixed_disc)
+
+    // join with quantities, add sec_id, sess_id,
+    val result = for {
+      prices_with_discounts <- prices_with_discounts
+    } yield joinPriceDiscountWithQuantities(prices_with_discounts, (orders map {order => order.inventoryId -> order.quantity}).toMap)
+
+    result.map(records => {
+      CartPurchases.insertAll(records)
+    })
+  }
+
+  private def joinPriceDiscountWithQuantities(prices_with_discounts: Seq[Cart.this.PriceDiscountPartialResult], quantityOfInventoryId: Map[Int, Int]) =
+    prices_with_discounts map {
+      case PriceDiscountPartialResult(s_id, i_id, i_price, i_min_price, i_fixed_disc) =>
+        CartPurchases.newRecord
+          .withCellContent(CartPurchases.sectionId)(s_id)
+          .withCellContent(CartPurchases.sessionId)(currentSessionId)
+          .withCellContent(CartPurchases.inventoryId)(i_id)
+          .withCellContent(CartPurchases.quantity)(quantityOfInventoryId.getOrElse(i_id, 0))
+          .withCellContent(CartPurchases.price)(i_price)
+          .withCellContent(CartPurchases.minPrice)(i_min_price)
+          .withCellContent(CartPurchases.fixedDiscount)(i_fixed_disc)
+          .build()
+    }
+
+  case class PriceDiscountPartialResult(s_id: Int, i_id: Int, i_price: Double, i_min_price: Double, i_fixed_disc: Double)
+  private def joinPriceAndDisc(ppr: Seq[Cart.this.PricePartialResult], dpr: Seq[Cart.this.DiscountsPartialResult]) = {
+    val discount_for_i_id = (dpr map {
+      a => a.i_id -> a.i_fixed_disc
+    }).toMap
```

Comment:
**Style - this is subjective**

Use dot-notation for getting rid of parenthesis:

```scala
val discount_for_i_id = dpr.map { a =>
  a.i_id -> a.i_fixed_disc
}.toMap
```